### PR TITLE
Refactor apply_tooltips

### LINKED ISSUE #8685: Not all the requested fieds are not shown in a tooltip
- Browsers: Chrome 69.0.3497.100, Firefox 62.0.3 (64-bit), EDGE 38.14393.2068.0)
- Operating system:  Windows 10
- Your databases, MongoDB
- Metabase version: 0.30.4
- Metabase hosting environment: Docker
- Metabase internal database: H2 (default)

**Steps to repro:**
1. Request 4 fields from your mongodb table. In my case I had the next query:  
`[{"$project":{"_id": 1, "name":"$test_name", "TPS":"$tps", "Status": "$test_status"}}]`
2. Create a bar chart like in attached screenshot: 
settings
3. Hover mouse over the several bars

**Actual results:**
Test name is not shown in a tooltip for several bars.  
Bar with all the requested fields: 
testname
Tooltips without name field:
no test name
no test name 2

__Expected results:__
Tooltip for the every bar should display the test name

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### LINKED ISSUE #9361: Stacked Bar Chart Drill In Incorrect
When a bar chart is stacked (either normal or percentage wise) the drill in sometimes does not produce the right query filter when creating a new question.

In the below bar chart highlighting the red "FAILURE" on January 31 and attempting to view the records generates a question with the filter being "SUCCESS" instead of "FAILURE":
image

## Steps to reproduce:
1. Create the following query
image

2. Change the visualization to be a stacked bar chart with date on the bottom
image

3. Drill in to the top right bar (Aerodynamic Leather Toucan) by clicking "view this order"
image

The generated question uses a filter of Product > Title = "Aerodynamic Bronze Hat". It should be "Aerodynamic Leather Toucan".
image

Your browser and the version: 71.0.3578.98 (Official Build) (64-bit)
Your operating system: Windows 7
Your databases: Oracle 12.2
Metabase version: 0.31.2
Metabase hosting environment: RHEL 7
Metabase internal database: MySQL

### PR DESCRIPTION
**Pull request by paulrosenzweig**
Resolves #9361
Resolves #8685

The core issue in #9361 was that we use different logic for data lookup with tooltips on hover vs drill-in on click. That led to the issue where the correct data was displayed in a tooltip, but on drill in incorrect data was used. This PR unifies the data-lookup logic for onClick and onHover.

Prior to this consolidation there were also some issues where tooltips wouldn't show all available data (#8685). In redoing how we find the data for a click/hover event, this PR also fixes that bug.

### What's our preferred tooltip key ordering?

Some of the ordering got jostled in the refactor. Which of these is better?

image
image
(The difference is date formatting isn't part of the change)

                                                                           
### Is adding a breakout action ok? 

This PR also added another action to some bar charts. I think this is fine and it probably should have been there before, but I thought I should check that there's no context I'm missing.

#### Before
image
#### After
image

### Dropping an extra breakout column reference?
Another unintentional change I noticed was dropping an extra breakout column reference in scatter plot tooltips. I can't imaging the old behavior was intentional, but I wanted to call out an unrelated behavior update.

#### Before
image
#### After
image

### CODE REVIEW

**Review by tlrobinson — COMMENTED** [review]
A couple questions/comments.

Glad to have this consolidated and not dependent on DOM element ordering.

File: frontend/src/metabase/visualizations/lib/apply_tooltips.js [hunk 1]
```diff
@@ -1,286 +1,238 @@
...
+          // 2. if not, we format the key using the column info
+          // this catches values like years that don't parse correctly above
+          formatValue(key, { column: rawCols[0] }) === String(x)
+        : // otherwise, we just check if the string value matches
+          // e.g. String("123") === String(123)
+          String(x) === String(key),
+    );
+
+    // try to get row from _origin but fall back to the row we already have
+    const { _origin: { row: rawRow = row } = {} } = row;
```
[the remark this answers is not in the document]
Comment (paulrosenzweig) [reply frontend/src/metabase/visualizations/lib/apply_tooltips.js]: 👍 the destructure-less version is definitely more readable. It makes the `row` fallback more clear. I'll swap it out.

File: frontend/src/metabase/visualizations/lib/apply_tooltips.js [hunk 2]
```diff
@@ -1,286 +1,238 @@
...
-      clicked.dimensions.push({
-        value: card._breakoutValue,
-        column: card._breakoutColumn,
+    const seriesData = series[seriesIndex].data || {};
+    const rawCols = seriesData._rawCols || cols;
+    const { key } = d.data;
+
+    // We look through the rows to match up they key in d.data to the x value
+    // from some row.
+    const row = seriesData.rows.find(([x]) =>
```
[the remark this answers is not in the document]
Comment (paulrosenzweig) [reply frontend/src/metabase/visualizations/lib/apply_tooltips.js]: I _think_ so. Scatter plots have duplicate x-values within a series, but those are handled separately. Any edge cases you can think of?

File: frontend/src/metabase/visualizations/lib/apply_tooltips.js [hunk 3]
```diff
@@ -1,286 +1,238 @@
...
+      moment.isMoment(key)
+        ? // for dates, we check two things:
+          // 1. does parsing x produce an equivalent moment value?
+          key.isSame(moment(x)) ||
+          // 2. if not, we format the key using the column info
+          // this catches values like years that don't parse correctly above
+          formatValue(key, { column: rawCols[0] }) === String(x)
+        : // otherwise, we just check if the string value matches
+          // e.g. String("123") === String(123)
+          String(x) === String(key),
```
Comment (walterl) [inline frontend/src/metabase/visualizations/lib/apply_tooltips.js]: Does this hold for all types that `x` could be?

File: frontend/src/metabase/visualizations/lib/apply_tooltips.js [hunk 4]
```diff
@@ -1,286 +1,238 @@
...
-            element: isArea ? null : this,
-            event: isArea ? d3.event : null,
-            data: data.length > 0 ? data : null,
-          });
+        .on("mousemove", function(d) {
+          const hovered = getClickHoverHelper(this, d);
+          onHoverChange(hovered);
         })
         .on("mouseleave", function() {
           if (!onHoverChange) {
```
Comment (walterl) [inline frontend/src/metabase/visualizations/lib/apply_tooltips.js]: This isn't part of your changes, but can this test ever be satisfied, given that it's enclosed in an `if (onHoverChange)`-block?

**Review by walterl — COMMENTED** [review]
I don't feel very confident in my review. Maybe it's better to get another set of eyes on this...?

**Review by tlrobinson — APPROVED** [review]
LGTM

**Pull request merged**